### ISSUE TITLE
Debugger: my array shown empty by debugger when there is a value in it.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12253]
Migrated from [rt.perl.org#114018](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114018) (status was 'resolved')

Searchable as RT114018$

**Comment by p5pRT** [comment]
### From trey.bianchini@gmail.com

#### Created by trey.bianchini@gmail.com




when an arrayref is declared and conditionally initialized 
my $ss=$args\{a\} if \( $args\{a\} \);
and then a real value is assigned to it the debugger  won't show it the value "x $ss"
declaring like
my $ss;
$ss=$args\{a\} if \( $args\{a\} \);
solves the problem


here is a sample module and test and debug session to replicate&#8203;:


$cat Showbug\.t

use strict;
use Showbug;

my $sf;
\#my $ts='aaa|aab\-test\,e';
my $ts='junk';
Showbug&#8203;::show\_bug\($sf\,ts=>$ts\);
Showbug&#8203;::no\_bug\($sf\,ts=>$ts\);


\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#END


$cat \./Showbug\.pm
package Showbug;
use strict;



sub assemble\_server\_stack \{

&nbsp;   my @&#8203;stack;
&nbsp;   push @&#8203;stack\, '1';
&nbsp;   push @&#8203;stack\, '2';
&nbsp;   return \\@&#8203;stack;

\}


sub show\_bug \{
&nbsp;    my \( $sf\, %args \) = @&#8203;\_;

&nbsp;    my $ss = $args\{server\_stack\} if \( $args\{server\_stack\} \);
&nbsp;    my $ts=$args\{ts\};

&nbsp;    if \( $ts \) \{
&nbsp;        $ss = assemble\_server\_stack\(ts=>$ts\);
\#        $DB&#8203;::single=1;
&nbsp;        my $crap=1;
&nbsp;        if \( $ss \) \{
&nbsp;           print @&#8203;\{$ss\} \. "\\n";
&nbsp;         \} else \{
&nbsp;           print "nope\\n";
&nbsp;        \}
&nbsp;    \}

\}

sub no\_bug \{
&nbsp;    my \( $sf\, %args \) = @&#8203;\_;

&nbsp;    my $ss;
&nbsp;    $ss = $args\{server\_stack\} if \( $args\{server\_stack\} \);
&nbsp;    my $ts=$args\{ts\};

&nbsp;    if \( $ts \) \{
&nbsp;        $ss = assemble\_server\_stack\(ts=>$ts\);
\#        $DB&#8203;::single=1;
&nbsp;        my $crap=1;
&nbsp;        if \( $ss \) \{
&nbsp;           print @&#8203;\{$ss\} \. "\\n";
&nbsp;         \} else \{
&nbsp;           print "nope\\n";
&nbsp;        \}
&nbsp;    \}
\}


1;

\#\#\#\#\#\#\#\#\#\#\#\#END


\#Debugging session showing bug
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#

$perl \-v


This is perl 5\, version 14\, subversion 2 \(v5\.14\.2\) built for x86\_64\-linux

Copyright 1987\-2011\, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License\, which may be found in the Perl 5 source kit\.

Complete documentation for Perl\, including FAQ lists\, should be found on
this system using "man perl" or "perldoc perl"\.  If you have access to the
Internet\, point your browser at http&#8203;://www\.perl\.org/\, the Perl Home Page\.


$perl \-d Showbug\.t

Loading DB routines from perl5db\.pl version 1\.33
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(Showbug\.t&#8203;:4\)&#8203;:    my $sf;
&nbsp;                                                                                                                     DB\<1> n
main&#8203;::\(Showbug\.t&#8203;:6\)&#8203;:    my $ts='junk';
&nbsp;                                                                                                                     DB\<1> n
main&#8203;::\(Showbug\.t&#8203;:7\)&#8203;:    Showbug&#8203;::show\_bug\($sf\,ts=>$ts\);
&nbsp;                                                                                                                     DB\<1> s
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:17\)&#8203;:           my \( $sf\, %args \) = @&#8203;\_;
&nbsp;                                                                                                                     DB\<1> n
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:19\)&#8203;:           my $ss = $args\{server\_stack\} if \( $args\{server\_stack\} \);
&nbsp;                                                                                                                     DB\<1> n
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:20\)&#8203;:           my $ts=$args\{ts\};
&nbsp;                                                                                                                     DB\<1> n
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:22\)&#8203;:           if \( $ts \) \{
&nbsp;                                                                                                                     DB\<1> n
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:23\)&#8203;:               $ss = assemble\_server\_stack\(ts=>$ts\);
&nbsp;                                                                                                                     DB\<1> n
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#                                                                      \#here is the the bug happening
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:25\)&#8203;:               my $crap=1;                                                          DB\<1> x $ss
0  undef
&nbsp;                                                                                                                     DB\<2> n
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:26\)&#8203;:               if \( $ss \) \{
&nbsp;                                                                                                                     DB\<2> n
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:27\)&#8203;:                  print @&#8203;\{$ss\} \. "\\n";
&nbsp;                                                                                                                     DB\<2> n
Showbug&#8203;::show\_bug\(Showbug\.pm&#8203;:27\)&#8203;:                  print @&#8203;\{$ss\} \. "\\n";
&nbsp;                                                                                                                     DB\<2> n
2
main&#8203;::\(Showbug\.t&#8203;:8\)&#8203;:    Showbug&#8203;::no\_bug\($sf\,ts=>$ts\);
&nbsp;                                                                                                                     DB\<2> s
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:36\)&#8203;:     my \( $sf\, %args \) = @&#8203;\_;
&nbsp;                                                                                                                     DB\<2> n
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:38\)&#8203;:     my $ss;
&nbsp;                                                                                                                     DB\<2> n
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:39\)&#8203;:     $ss = $args\{server\_stack\} if \( $args\{server\_stack\} \);
&nbsp;                                                                                                                     DB\<2> n
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:40\)&#8203;:     my $ts=$args\{ts\};
&nbsp;                                                                                                                     DB\<2> n
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:42\)&#8203;:     if \( $ts \) \{
&nbsp;                                                                                                                     DB\<2> n
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:43\)&#8203;:         $ss = assemble\_server\_stack\(ts=>$ts\);
&nbsp;                                                                                                                     DB\<3> n
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#                                                                    \#here is the the normal behavior
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:45\)&#8203;:         my $crap=1;
&nbsp;                                                                                                                     DB\<3> x $ss
0  ARRAY\(0x144fef8\)
&nbsp;   0  1
&nbsp;   1  2
&nbsp;                                                                                                                     DB\<4> n
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:46\)&#8203;:         if \( $ss \) \{
&nbsp;                                                                                                                     DB\<4> n
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:47\)&#8203;:            print @&#8203;\{$ss\} \. "\\n";
&nbsp;                                                                                                                     DB\<4> n
Showbug&#8203;::no\_bug\(Showbug\.pm&#8203;:47\)&#8203;:            print @&#8203;\{$ss\} \. "\\n";
&nbsp;                                                                                                                     DB\<4> n
2
Debugged program terminated\.  Use q to quit or R to restart\,
&nbsp;  use o inhibit\_exit to avoid stopping after program termination\,
&nbsp;  h q\, h R or h o to get additional info\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.14.2:

Configured by tbianchini at Fri Mar  2 11:29:51 CST 2012.

Summary of my perl5 (revision 5 version 14 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.32-5-amd64, archname=x86_64-linux
    uname='linux prg-stg-salesforce.prg.com 2.6.32-5-amd64 #1 smp thu nov 3 03:41:26 utc 2011 x86_64 gnulinux '
    config_args='-de -Dprefix=/home/tbianchini/perl5/perlbrew/perls/perl-5.14.2'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.5', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.14.2:
    /home/tbianchini/src/prg-salesforce/lib
    /home/tbianchini/src/prg-core/lib
    /home/tbianchini/src/prg-orderexpress/lib
    /home/tbianchini/src/net-bullfinch/lib
    /home/tbianchini/src/prg-config/lib
    /home/tbianchini/perl5/perlbrew/perls/perl-5.14.2/lib/site_perl/5.14.2/x86_64-linux
    /home/tbianchini/perl5/perlbrew/perls/perl-5.14.2/lib/site_perl/5.14.2
    /home/tbianchini/perl5/perlbrew/perls/perl-5.14.2/lib/5.14.2/x86_64-linux
    /home/tbianchini/perl5/perlbrew/perls/perl-5.14.2/lib/5.14.2
    .


Environment for perl 5.14.2:
    HOME=/home/tbianchini
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/home/jhannah/src/oracle/instantclient_11_2
    LOGDIR (unset)
    PATH=/home/tbianchini/perl5/perlbrew/bin:/home/tbianchini/bin:/home/tbianchini/perl5/perlbrew/bin:/home/tbianchini/perl5/perlbrew/perls/perl-5.14.2/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
    PERL5LIB=/home/tbianchini/src/prg-salesforce/lib:/home/tbianchini/src/prg-core/lib:/home/tbianchini/src/prg-orderexpress/lib:/home/tbianchini/src/net-bullfinch/lib:/home/tbianchini/src/prg-config/lib
    PERLBREW_BASHRC_VERSION=0.41
    PERLBREW_HOME=/home/tbianchini/.perlbrew
    PERLBREW_MANPATH=/home/tbianchini/perl5/perlbrew/perls/perl-5.14.2/man
    PERLBREW_PATH=/home/tbianchini/perl5/perlbrew/bin:/home/tbianchini/perl5/perlbrew/perls/perl-5.14.2/bin
    PERLBREW_PERL=perl-5.14.2
    PERLBREW_ROOT=/home/tbianchini/perl5/perlbrew
    PERLBREW_VERSION=0.41
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in 7ef308308\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'resolved'
